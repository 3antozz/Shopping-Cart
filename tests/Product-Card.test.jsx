import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ProductCard from "../src/Components/Product-Card/Product-Card.jsx";
import { MemoryRouter } from "react-router-dom";

describe("Product Card component", () => {
    let onSubmit;
    let handlePopup;
    beforeEach(() => {
        onSubmit = vi.fn();
        handlePopup = vi.fn();
        const product = {
                id: 1,
                title: "test",
                price: 69,
                category: "...",
                description: "...",
                image: "...",
            }
        render(
            <MemoryRouter initialEntries={['/shop']}>
                <ProductCard 
                product={product}
                onSubmit={onSubmit}
                handlePopup={handlePopup} />
            </MemoryRouter>
        );
    });
    it("card container is rendered", () => {
        expect(screen.getByTestId(/container/i)).toBeInTheDocument();
    })
    it("navigates to product link", () => {
        expect(screen.getByText(/test/i)).toHaveAttribute("href", "/products/1");
    });
    it("should call prop functions when clicked", async () => {
        const user = userEvent.setup();
        const cartButton = screen.getByRole("button", {name: /Add to Cart/i});

        await user.click(cartButton);
        expect(handlePopup).toBeCalled();
        expect(onSubmit).toBeCalled();
    })
    it("should increment and decrement when clicked", async () =>{
        const user = userEvent.setup();
        const incrementButton = screen.getByRole("button", {name: "+"});
        const decrementButton = screen.getByRole("button", {name: "-"});

        const input = screen.getByDisplayValue("1");
        await user.click(incrementButton);
        expect(input.value).toBe("2");
        await user.click(decrementButton);
        expect(input.value).toBe("1");
    })
    it("matches snapshot", () => {
        const onSubmit = vi.fn();
        const handlePopup = vi.fn();
        const product = {
                id: 1,
                title: "test",
                price: 69,
                category: "...",
                description: "...",
                image: "...",
            }
        const {container} = render(
            <MemoryRouter>
                <ProductCard 
                    product={product}
                    onSubmit={onSubmit}
                    handlePopup={handlePopup}
                />
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });
});
