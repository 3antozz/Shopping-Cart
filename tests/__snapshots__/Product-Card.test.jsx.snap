// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Card component > matches snapshot 1`] = `
<div>
  <div
    class="_card_621a75"
    data-testid="container"
  >
    <a
      data-discover="true"
      href="/products/1"
    >
      <img
        alt="test"
        src="..."
      />
    </a>
    <a
      data-discover="true"
      href="/products/1"
    >
      test
    </a>
    <form
      class="_quantity_621a75"
    >
      <div
        class="_input_621a75"
      >
        <h4>
          69
          Â $
        </h4>
        <label
          for="quantity"
        />
        <button
          type="button"
        >
          -
        </button>
        <input
          id="quantity"
          max="100"
          min="1"
          type="number"
          value="1"
        />
        <button
          class="increment"
          type="button"
        >
          +
        </button>
      </div>
      <button
        type="submit"
      >
        Add to Cart
      </button>
    </form>
  </div>
</div>
`;
